### ISSUE TITLE
Invalid LeanCloud Counter request when page url contains character `&` 

### ISSUE THREAD

**Issue by EmmetZC** [theme-next/hexo-theme-next#1399]
If you delete this template, we may close your issue without conducting an investigation.
如果您删除此模版，我们可能会在不进行调查的情况下关闭您的 Issue。

Please check if your Issue fulfills the following requirements. This will guide you to resolve the issue quickly.
请务必确认您在提交此 Issue 前已经查看并完成了下列检查单中的所有事项。这会指导您快速地解决问题。

## Issue Checklist 

- [X] I am using [the latest](https://github.com/theme-next/hexo-theme-next/releases/latest) version of NexT.
- [X] I have reviewed [guidelines for NexT](https://github.com/theme-next/hexo-theme-next/blob/master/.github/CONTRIBUTING.md), especially the [Before Submitting An Issue](https://github.com/theme-next/hexo-theme-next/blob/master/.github/CONTRIBUTING.md#before-submitting-an-issue) part.
- [X] I have reviewed the latest Roadmap on GitHub and searched for current [issues](https://github.com/theme-next/hexo-theme-next/issues?utf8=%E2%9C%93&q=is%3Aissue), which does not help me.

***

## Expected behavior 
Get post visit count from leancloud

## Actual behavior 

- Links to demo site with this issue: https://emmetzc.github.io/2019/07/16/Diary&Dairy/
- Links to repository or source code of the blog: https://github.com/EmmetZC/EmmetZC.github.io/blob/89b8bb24efdf5e3b0741fbe47b69fa288b7ba6c8/2019/07/16/Diary%26Dairy/index.html#L404
In fact, I've fixed this issue by wrapping `JSON.stringify({ url })` into `encodeURIComponent()`, so you won't acturally see it. But I'm pretty sure that this problem happens because the template doesn't handle special characters, such as `&` in my example, properly.
Patch I used to fix this problem:
```
diff --git a/layout/_third-party/statistics/lean-analytics.swig b/layout/_third-party/statistics/lean-analytics.swig
index ac44730..6c5992b 100644
--- a/layout/_third-party/statistics/lean-analytics.swig
+++ b/layout/_third-party/statistics/lean-analytics.swig
@@ -10,7 +10,7 @@
       var url = decodeURI(visitors.getAttribute('id'));
       var title = visitors.dataset.flagTitle;

-      Counter('get', `/classes/Counter?where=${JSON.stringify({ url })}`)
+      Counter('get', `/classes/Counter?where=${encodeURIComponent(JSON.stringify({ url }))}`)
         .then(response => response.json())
         .then(({ results }) => {
           if (results.length > 0) {
@@ -55,7 +55,7 @@
         return decodeURI(element.getAttribute('id'));
       });

-      Counter('get', `/classes/Counter?where=${JSON.stringify({ url: { '$in': entries } })}`)
+      Counter('get', `/classes/Counter?where=${encodeURIComponent(JSON.stringify({ url: { '$in': entries } }))}`)
         .then(response => response.json())
         .then(({ results }) => {
           if (results.length === 0) {
```

## Steps to reproduce the behavior 
1. Use leancloud to do visit-count-statistics
2. Create a post which has `&` in its url
3. Visit the post and you'll see you failed fetching the visit-count because your query is not valid json.
4. Since visit-count-statistics won't work on `localhost` testing, you may need to deploy it to reproduce the problem.

By the way, the post produces this problem was written long ago, and it worked fine back then. That time I was using NeXT theme v7.2.0. Earlier today, I upgraded NeXT to the latest version and this problem newly appeared.

### Node.js and NPM Information

```
v12.16.1
6.13.4
```

### Package dependencies Information

```
├── @fortawesome/free-brands-svg-icons@5.9.0
├── @fortawesome/free-regular-svg-icons@5.9.0
├── @fortawesome/free-solid-svg-icons@5.9.0
├── babel-runtime@6.26.0
├── eslint@6.1.0
├── hexo@3.9.0
├── hexo-deployer-git@1. 0.0
├── hexo-fontawesome@2.0.0
├── hexo-generator-archive@0.1.5
├── hexo-generator-category@0.1.3
├── hexo-generator-index@0.2.1
├── hexo-generator-tag@0.2.0
├── hexo-leancloud-counter-security@1.4.0
├── hexo-renderer-ejs@0.3.1
├── hexo-renderer-marked@1.0.1
├── hexo-renderer-stylus@0.3.3
└── hexo-server@0.3.3
```

### Hexo Configuration

```yml
url: https://emmetzc.github.io/
root: /
permalink: :year/:month/:day/:title/
permalink_defaults:

new_post_name: :title.md # File name of new posts
default_layout: post
titlecase: false # Transform title into titlecase
external_link: true # Open external links in new tab
filename_case: 0
render_drafts: false
post_asset_folder: false
relative_link: false
future: true
highlight:
  enable: true
  line_number: true
  auto_detect: false
  tab_replace:
```

### NexT Configuration

```yml

```

**Comment by stevenjoezhang** [comment]
Please upgrade to Hexo@4. URL is encoded by default.

**Comment by stevenjoezhang** [comment]
Please upgrade to Hexo@4 https://github.com/hexojs/hexo/releases/tag/4.0.0
Currently you're using hexo@3.9.0 which does not encode URLs by default

**Comment by EmmetZC** [comment]
> Please upgrade to Hexo@4 https://github.com/hexojs/hexo/releases/tag/4.0.0
> Currently you're using hexo@3.9.0 which does not encode URLs by default
> 
> > <img alt="" width="48" src="https://avatars0.githubusercontent.com/u/6375567?s=400&amp;v=4"><img alt="" height="14" src="https://camo.githubusercontent.com/4908d9947135e4536ef61d61512ee575f869bf6f/68747470733a2f2f6769746875622e6769746875626173736574732e636f6d2f66617669636f6e2e69636f"> GitHub**[hexojs/hexo](https://github.com/hexojs/hexo)**A fast, simple & powerful blog framework, powered by Node.js. - hexojs/hexo

Upgated to Hexo@4.2.0 and still not working.
Hexo@4 encodes url by default for sure, but character `&` requires `encodeURIComponent()` to be encoded properly.

**Comment by stevenjoezhang** [comment]
Hexo doesn't seem to recommend using special symbols in post titles: https://github.com/hexojs/hexo/issues/4104
Anyway, I pushed the bugfix using `encodeURIComponent` to the master branch, you can update and try if this issue is solved. Hexo 4 is not backward compatible with URL encode changes, which makes these encode / decode problems tricky

**Comment by EmmetZC** [comment]
Yeah, right. Apologies for making this problem. Better not use special
characters again in the future.I've tried the latest NexT theme and it works
fine.Many thanks for your help.  





On Sun, Mar 8, 2020 4:42 AM, Mimi notifications@github.com  wrote:
Hexo doesn't seem to recommend using special symbols in post titles: 
hexojs/hexo#4104
Anyway, I pushed the bugfix using encodeURIComponent  to the master branch, you
can update and try if this issue is solved. Hexo 4 is not backward compatible
with URL encode changes, which makes these encode / decode problems tricky

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.



Best Wishes

(Chi Zhang)张驰
+86 13279345523
西安交通大学 电信学院
Xi'an Jiaotong University
The School of Electronic and Information Engineering
